File: 06ros/ws02/second_ws/src/use_cus_srv_pkg/src/server_cus_srv.cpp
//
// Created by zcb on 5/4/21.
//

#include <ros/ros.h>
#include <iostream>
#include <custom_srvs/NumOption.h>

bool callback(custom_srvs::NumOptionRequest &request, custom_srvs::NumOptionResponse &response) {
    double a = request.a;
    double b = request.b;
    std::string op = request.op;
    if (op == "+") {
        response.res = a + b;
    } else if (op == "-") {
        response.res = a - b;
    } else if (op == "*") {
        response.res = a * b;
    } else if (op == "/") {
        response.res = a / b;
    } else if (op == "%") {
        response.res = fmod(a, b);
    } else {
        return false;
    }
    return true;
}


int main(int argc, char **argv) {

    std::string nodeName = "server_node";
    ros::init(argc, argv, nodeName);

    ros::NodeHandle node;

    std::string serviceName = "/zcb01/service";
    ros::ServiceServer server = node.advertiseService(serviceName, callback);

    ros::spin();


    return 0;
}

File: 06ros/ws02/second_ws/src/use_cus_srv_pkg/scripts/server_cus_srv.py
#! /usr/bin/env python
# coding=utf8

import rospy
from custom_srvs.srv import NumOption, NumOptionRequest, NumOptionResponse


def callback(request):
    if not isinstance(request, NumOptionRequest):
        return
    a = request.a
    b = request.b
    op = request.op

    response = NumOptionResponse()
    if op == "+":
        response.res = a + b
    elif op == "-":
        response.res = a - b
    elif op == "*":
        response.res = a * b
    elif op == "/":
        response.res = a / b
    elif op == "%":
        response.res = a % b

    return response


if __name__ == '__main__':
    nodeName = "server_node"
    rospy.init_node(nodeName)

    serviceName = "/zcb01/service"
    server = rospy.Service(serviceName, NumOption, callback)  # server

    rospy.spin()
